monkey/monkey_island/cc/services/authentication_service/setup.py

```diff
@@ -50,3 +54,32 @@ def _build_authentication_facade(container: DIContainer, security: Security):
         token_parser,
         container.resolve(MongoOTPRepository),
     )
+
+
+def _register_event_handlers(container: DIContainer, authentication_facade: AuthenticationFacade):
+    agent_event_queue = container.resolve(IAgentEventQueue)
+    island_event_queue = container.resolve(IIslandEventQueue)
+
+    agent_event_queue.subscribe_type(
+        AgentShutdownEvent, unregister_agent_on_shutdown(authentication_facade)
+    )
+    island_event_queue.subscribe(
+        IslandEventTopic.AGENT_TIMED_OUT, unregister_agent_on_timeout(authentication_facade)
+    )
+
+
+class unregister_agent_on_shutdown:
+    def __init__(self, authentication_facade: AuthenticationFacade):
+        self._authentication_facade = authentication_facade
+
+    def __call__(self, event: AbstractAgentEvent):
+        agent_id = event.source
+        self._authentication_facade.remove_user(str(agent_id))
+
+
+class unregister_agent_on_timeout:
+    def __init__(self, authentication_facade: AuthenticationFacade):
+        self._authentication_facade = authentication_facade
+
+    def __call__(self, agent_id: AgentID):
+        self._authentication_facade.remove_user(str(agent_id))
```

Comment:
Small suggestion:

These could be one component with two methods. Those methods could be registered for their respective events. They could also share what little code there is.

Reply:
We can possibly get rid of `unregister_agent_on_shutdown` if we have the Agent user log out using `api/logout` before it shuts down.

Reply:
Never mind, authentication tokens may be left un-revoked in that case if the `AgentShutdownEvent` is sent but it couldn't log out after that for some reason.